Code/Framework/AzCore/AzCore/Jobs/Internal/JobManagerWorkStealing.cpp

```diff
@@ -77,7 +78,135 @@ Job* WorkQueue::TryStealFront()
 
     return nullptr;
 }
+#else // JOBMANAGER_LEGACY_WORKQUEUE
+struct StatusPayload
+{
+    StatusPayload(uint64_t status)
+    {
+        memcpy(this, &status, sizeof(status));
+    }
+
+    StatusPayload& operator=(uint64_t status)
+    {
+        memcpy(this, &status, sizeof(status));
+        return *this;
+    }
+
+    uint64_t pack() const
+    {
+        uint64_t out;
+        memcpy(&out, this, sizeof(out));
+        return out;
+    }
+
+    [[nodiscard]] bool empty() const
+    {
+        return head == tail;
+    }
+
+    // Typical circular buffer pattern where the head chases the tail. For two phase commits, tailReserve can run ahead of tail
+    // but can't bump into head. Note that for all operations here, we rely on well-defined wrapping behavior of unsigned ints.
+    [[nodiscard]] bool canEnqueue() const
+    {
+        return tailReserve != head - 1;
+    }
+
+    void reserve()
+    {
+        ++tailReserve;
+    }
+
+    void commit()
+    {
+        ++tail;
+    }
+
+    void acquire()
+    {
+        ++head;
+    }
+
+    uint16_t head;
+    uint16_t tail;
+    uint16_t tailReserve;
+    uint16_t unused;
+};
+
+void JobQueue::Enqueue(Job* job)
+{
+    uint8_t priority = job->GetPriority();
+    uint64_t previousStatus = m_status[priority].load(AZStd::memory_order_acquire);
+
+    while (true)
+    {
+        StatusPayload status{ previousStatus };
+
+        // Enqueuing is done in two phases because we cannot atomically write the job to the slot we reserve
+        // and simulataneously publish the fact that the slot is now available.
+        if (status.canEnqueue())
+        {
+            // Try to reserve a slot
+            status.reserve();
+            uint64_t reserveStatus = status.pack();
+            if (m_status[priority].compare_exchange_weak(previousStatus, reserveStatus))
+            {
```

Comment:
Wouldn't it be better to move the `while(true)` under `if(status.canEnqueue())`? The status is never refeteched so stays the same and the alternative path where the queue is full doesn't seem to require looping. If the main loop is setup this way to prevent a failure to commit to recheck if the queue is available, then the status needs to be refetched.

Reply:
the status changes each time the compare exchange occurs and fails due to spurious wakeup.